File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {
	static int n , m;
	static int[][] board;
	static boolean[][] visited;
	static ArrayDeque<int[]> q = new ArrayDeque<>();
	static int[] dx = {1,-1,0,0};
	static int[] dy = {0,0,1,-1};
	static int ans;
	
	static int[][] copy(int[][] b){
		int[][] copied = new int[n][m];
		for(int i = 0 ; i < n ; i++) {
			for(int j = 0 ; j < m ; j++) {
				copied[i][j] = b[i][j];
			}
		}
		return copied;
		
	}
	static void infect(int x, int y, int[][] b){
		
		
		q.add(new int[] {x,y});
		visited[x][y] = true;
		while(!q.isEmpty()) {
			int[] tmp = q.pollFirst();
			int xx = tmp[0];
			int yy = tmp[1];
			for(int i = 0 ; i < 4 ; i++) {
				int nx = xx + dx[i];
				int ny = yy + dy[i];
				if(0<=nx && nx < n && 0<= ny && ny < m && b[nx][ny] == 0) {
					b[nx][ny] = 2;
					visited[nx][ny] = true;
					q.addLast(new int[] {nx,ny});
				}
			}
		}
		
	}
	
	static int getArea(int[][] b) {
		int result = 0;
		for(int i = 0 ; i < n ; i++) {
			for(int j = 0 ; j < m ; j++) {
				if(b[i][j] == 0) result += 1;
			}
		}
		return result;
	}
	static void dfs(int x, int y, int[][] b,int cnt) {
		if(cnt == 3) {
			int[][] copied = copy(b);
			visited = new boolean[n][m];
			for(int i = 0 ; i < n ; i++) {
				for(int j = 0 ; j < m ; j++) {
					if (b[i][j] == 2) {
						if (!visited[i][j])
							infect(i,j,copied);
					}
				}
			}
			ans = Math.max(getArea(copied),ans);
			return;
			
		}
		if(x == n) {
			return;
		}
		if(y == m) {
			dfs(x+1,0,b,cnt);
			return;
		}
		if(b[x][y] == 0) {
			b[x][y] = 1;
			dfs(x,y+1,b,cnt+1);
			b[x][y] = 0;
		}
		dfs(x,y+1,b,cnt);
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		board = new int[n][m];
		for(int i = 0; i < n ; i++){
			st = new StringTokenizer(br.readLine());
			for(int j = 0 ; j < m ; j++) {
				board[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		dfs(0,0,board,0);
		System.out.println(ans);
	}

}

File: 백준/Gold/14502. 연구소/연구소.py
import copy
from collections import deque
n,m = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(n)]
ans = 0
# dfs로 기둥(1)을 지도상에 세우고 기둥 갯수 == 3일 때, 바이러스가 퍼지는 함수 써서 0의 갯수 max값 찾기
tmp = []
def dfs(N):
	global board,ans,tmp
	if N == 3 :
		ans = max(ans,virus(board))
		
		return
	for i in range(len(board)):
		for j in range(len(board[0])):
			if board[i][j] == 0 and [i,j] not in tmp :
				if tmp :
					if (tmp[-1][0] * len(board)) + tmp[-1][1] > (i * len(board)) + j :
						continue
				tmp.append([i,j])
				board[i][j] = 1
				dfs(N+1)
				tmp.pop()
				board[i][j] = 0
a = []
def virus(board):
	global a
	b = copy.deepcopy(board)
	s = 0
	q = deque()
	dx = [1,-1,0,0]
	dy = [0,0,1,-1]
	for i in range(len(board)):
		for j in range(len(board[0])):
			if b[i][j] == 2 :
				q.append([i,j])
				while q:
					x,y = q.popleft()
					for k in range(4):
						nx = x + dx[k]
						ny = y + dy[k]
						if 0<=nx< len(b) and 0<= ny < len(b[0]) and b[nx][ny] == 0 :
							q.append([nx,ny])
							b[nx][ny] = 2
							
	for bb in b :
		s += bb.count(0)		
	return s
	

dfs(0)
print(ans)
